bokeh/models/contour_renderer.py

```diff
@@ -0,0 +1,120 @@
+# General API
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+_DEFAULT_CONTOUR_LINE_RENDERER = lambda: GlyphRenderer(
+    glyph=MultiLine(), data_source=ColumnDataSource(data=dict())
+)
+
+_DEFAULT_CONTOUR_FILL_RENDERER = lambda: GlyphRenderer(
+    glyph=MultiPolygons(), data_source=ColumnDataSource(data=dict())
+)
+
+class ContourRenderer(DataRenderer):
+    '''
+    Renderer for contour plots composed of filled polygons and/or lines.
+    Rather than create these manually it is usually better to use
+    ``figure.contour`` instead.
+    '''
+
+    # explicit __init__ to support Init signatures
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    line_renderer = Instance(GlyphRenderer, default=_DEFAULT_CONTOUR_LINE_RENDERER, help="""
+    Glyph renderer used for contour lines.
+    """)
+
+    fill_renderer = Instance(GlyphRenderer, default=_DEFAULT_CONTOUR_FILL_RENDERER, help="""
+    Glyph renderer used for filled contour polygons.
+    """)
+
+    levels = Seq(Float, default=[], help="""
+    Levels at which the contours are calculated.
+    """)
+
+    def __setattr__(self, name: str, value: Unknown) -> None:
+        if name == "data":
+            fill_data, line_data = value
+
+            if fill_data:
+                # Copy fill and hatch properties from old to new data source
+                old_fill_data = self.fill_renderer.data_source.data
+                for name in old_fill_data.keys():
+                    if name not in ("xs", "ys", "lower_levels", "upper_levels"):
+                        fill_data[name] = old_fill_data[name]
+                self.fill_renderer.data_source.data = fill_data
+            else:
+                self.fill_renderer.data_source.data = dict(xs=[], ys=[], lower_levels=[], upper_levels=[])
+
+            if line_data:
+                # Copy line properties from old to new data source
+                old_line_data = self.line_renderer.data_source.data
+                for name in old_line_data.keys():
+                    if name not in ("xs", "ys", "levels"):
+                        line_data[name] = old_line_data[name]
+                self.line_renderer.data_source.data = line_data
+            else:
+                self.line_renderer.data_source.data = dict(xs=[], ys=[], levels=[])
+        else:
+            super().__setattr__(name, value)
+
+    def color_bar(self) -> ContourColorBar:
+        # Should accept other kwargs and pass them through?
+        return ContourColorBar(
+            fill_renderer=self.fill_renderer,
+            line_renderer=self.line_renderer,
+            levels=self.levels,
+            ticker=FixedTicker(ticks=self.levels),
+        )
```

Comment:
No, I can't image a scenario where it would be called more than once.

I wanted an easy way to create a `ContourColorBar` without the user having to worry about how it works (referencing the line and fill renderers). I thought a member function of `ContourRenderer` was a better approach to a standalone function, but I appreciate that this isn't typical in Bokeh so I am happy to do it some other way.